crates/router/src/core/payment_methods/cards.rs

```diff
@@ -1394,13 +1727,110 @@ pub async fn add_payment_method(
+pub async fn vault_payment_method(
+    state: &routes::SessionState,
+    pmd: &api::PaymentMethodCreateData,
+    merchant_account: &domain::MerchantAccount,
+    key_store: &domain::MerchantKeyStore,
+    customer_id: &id_type::CustomerId,
+) -> errors::RouterResult<payment_methods::VaultAddResp> {
+    let db = &*state.store;
+    let merchant_id = merchant_account.get_id();
+
+    // get fingerprint_id from locker
+    let fingerprint_id_from_locker =
+        get_fingerprint_id_from_locker(state, merchant_account, customer_id, pmd).await?;
+
+    // throw back error if payment method is duplicated
+    if db
+        .find_payment_method_by_fingerprint_id(
+            &(state.into()),
+            key_store,
+            &fingerprint_id_from_locker,
+        )
+        .await
+        .ok()
+        .is_some()
+    {
+        return Err(report!(errors::ApiErrorResponse::DuplicatePaymentMethod)
+            .attach_printable("Cannot vault duplicate payment method"));
+    }
+
+    let resp_from_locker =
+        vault_payment_method_in_locker(state, merchant_account, customer_id, pmd).await?;
+
+    Ok(resp_from_locker)
+}
+
+#[cfg(all(feature = "v2", feature = "payment_methods_v2"))]
+#[instrument(skip_all)]
+pub async fn get_fingerprint_id_from_locker(
+    state: &routes::SessionState,
+    merchant_account: &domain::MerchantAccount,
+    customer_id: &id_type::CustomerId,
+    pmd: &api::PaymentMethodCreateData,
+) -> errors::RouterResult<String> {
+    let card = match pmd {
+        api::PaymentMethodCreateData::Card(card) => Some(card.clone()),
+    }
+    .ok_or(errors::ApiErrorResponse::NotSupported {
+        message: "Payment method not supported".to_string(),
+    })?;
+
+    let key = card.card_number.clone().to_string();
+    let data = serde_json::to_value(card)
+        .change_context(errors::ApiErrorResponse::InternalServerError)
+        .attach_printable("Failed to encode CardDetail to value")?
+        .to_string();
+
+    let payload = payment_methods::VaultFingerprintGenericReq { key, data };
+
+    let resp = call_to_vault(state, &payload, customer_id)
+        .await
+        .change_context(errors::ApiErrorResponse::InternalServerError)
+        .attach_printable("Failed to get response from locker")?;
+
+    let fingerprint_resp: payment_methods::VaultFingerprintGenericResp = resp
+        .parse_struct("VaultFingerprintGenericResp")
+        .change_context(errors::ApiErrorResponse::InternalServerError)
+        .attach_printable("Failed to parse data into VaultFingerprintGenericResp")?;
+
+    Ok(fingerprint_resp.fingerprint_id)
+}
+
+#[cfg(all(feature = "v2", feature = "payment_methods_v2"))]
+#[instrument(skip_all)]
+pub async fn vault_payment_method_in_locker(
+    state: &routes::SessionState,
+    merchant_account: &domain::MerchantAccount,
+    customer_id: &id_type::CustomerId,
+    pmd: &api::PaymentMethodCreateData,
+) -> errors::RouterResult<payment_methods::VaultAddResp> {
+    let card = match pmd {
```

Comment:
This can't be a specific function of card here, you need to implement a trait which should get key for any payment type

Reply:
Didn't implement a trait as there was only one pm that's accepted in the flow as of now.